### PR TITLE
OSSMDOC-550: Document new WasmPlugin API.

### PR DESCRIPTION
**Pull request by JStickler**
[OSSMDOC-550](https://issues.redhat.com//browse/OSSMDOC-550): Document new WasmPlugin API.

OCP versions 4.6 - 4.11

This PR:
- Breaks the existing extensions assembly into ServiceMeshExtension modules
- Adds modules for the new WasmPlugin extension

Link to final docs preview: http://file.bos.redhat.com/jstickle/OSSMDOC-550/service_mesh/v2x/ossm-extensions.html 

<strike>Link to docs preview: https://deploy-preview-45111--osdocs.netlify.app/openshift-enterprise/latest/service_mesh/v2x/ossm-extensions</strike>

Eng review - dgn, jwendell
QE review - skondkar
Peer review - jldohmann

### CODE REVIEW

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+|AUTHZ
+|Insert plugin before Istio authorization filters and after Istio authentication filters.
+
+|STATS
+|Insert plugin before Istio stats filters and after Istio authorization filters.
+|===
+
+
+`Int64Value` specifies the wrapper message for int64. The JSON representation for Int64Value is a JSON string.
```
Comment (dgn) [inline modules/ossm-extensions-ref-wasmplugin.adoc:178]: not sure what this is?

File: modules/ossm-extensions-container-format.adoc [hunk 2]
```diff
@@ -0,0 +1,57 @@
+////
+This module included in the following assemblies:
+*service_mesh_/v2x/ossm-extensions.adoc
+////
+:_content-type: REFERENCE
+[id="ossm-extensions-format_{context}"]
+= WASM container format
```
Comment (dgn) [inline modules/ossm-extensions-container-format.adoc:7]: note that this format is only for `ServiceMeshExtension` (ie the legacy API). The new WasmPlugin API uses this format: https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/how-to-build-oci-images.md

the big change is that the `manifest.yaml` does not exist in that format, and the .wasm file always needs to be called `plugin.wasm`

**Review by dgn — APPROVED** [review]
lgtm except for the container image format

File: modules/ossm-extensions-container-format.adoc (hunk 2 above)
Comment (JStickler) [reply modules/ossm-extensions-container-format.adoc:7]: "the .wasm file always needs to be called plugin.wasm"

Does that mean that users are restricted to only being able to create a single plugin?

Comment (dgn) [reply modules/ossm-extensions-container-format.adoc:7]: within a single container: yes. but that has been the case with ServiceMeshExtension as well

**Comment by JStickler** [comment]
@dgn I revised the WasmPlugin container format topic.  I could use a second review for that topic please.

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+////
+:_content-type: REFERENCE
+[id="ossm-wasm-ref-wasmplugin_{context}"]
+= WasmPlugin API reference
+
+The WasmPlugins API provides a mechanism to extend the functionality provided by the Istio proxy through WebAssembly filters.
+
+You can deploy multiple WasmPlugins. The order of execution (as part of Envoy’s filter chain) is determined by the `phase` and `priority` settings, allowing the configuration of complex interactions between user-supplied WasmPlugins and Istio’s internal filters.
+
+In the following example, an authentication filter is deployed to the ingress-gateway that implements an OpenID flow and populates the `Authorization` header with a JSON Web Token (JWT) to be consumed by Istio authentication.
```
Comment (jwendell) [inline modules/ossm-extensions-ref-wasmplugin.adoc:13]: I found this sentence a bit hard to follow. Inverting it a little bit makes it easier for me. Something like:

In the following example, an authentication filter that implements an OpenID flow and populates the `Authorization` header with a JSON Web Token (JWT) to be consumed by Istio authentication is deployed to the ingress gateway. The WasmPlugin file lives in the proxy sidecar filesystem. Note the field `url`.

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+|Determines where in the filter chain this `WasmPlugin` is injected.
+|No
+
+|spec.priority
+|`int64`
+|Determines ordering of `WasmPlugins` in the same `phase`. When multiple `WasmPlugins` are applied to the same workload in the same `phase`, they will be applied by priority, in descending order. If `priority` is not set, or two `WasmPlugins` exist with the same
+value, the ordering will be deterministically derived from name and namespace of the `WasmPlugins`. Defaults to `0`.
+|No
+
+|spec.pluginName:
```
Comment (jwendell) [inline modules/ossm-extensions-ref-wasmplugin.adoc:100]: extra `:`

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+|string
+|The plugin name used in the Envoy configuration. Some .wasm modules might require this value to select the Wasm plugin to execute.
```
Comment (jwendell) [inline modules/ossm-extensions-ref-wasmplugin.adoc:102]: Suggestion: replace `.wasm` with `Wasm` or `WASM`.

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+|The configuration that will be passed on to the plugin.
+|No
+
+|spec.pluginConfig.verificationKey
+|string
+|Public key that will be used to verify signatures of signed OCI images or Wasm modules. Must be supplied in PEM format.
+|No
+|===
+
+`WorkloadSelector` specifies the criteria used to determine if a policy can be applied to a proxy. The matching criteria includes the metadata associated with a proxy, workload instance information such as labels attached to the pod/VM, or any other information that the proxy provides to Istio during the initial handshake. If multiple conditions are specified, all conditions need to match in order for the workload instance to be selected. Currently, only label based selection mechanism is supported.
```
Comment (jwendell) [inline modules/ossm-extensions-ref-wasmplugin.adoc:116]: Replace `policy` with `filter`.

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+|Yes
+|===
+
+`PullPolicy` specifies the pull behavior to be applied when fetching an OCI image.
+
+.PullPolicy
+[options="header"]
+[cols="a, a"]
+|===
+| Field | Description
```
Comment (jwendell) [inline modules/ossm-extensions-ref-wasmplugin.adoc:135]: Replace `Field` with `Value`

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+|UNSPECIFIED_POLICY
```
Comment (jwendell) [inline modules/ossm-extensions-ref-wasmplugin.adoc:136]: Replace this with `<empty>`

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+|Defaults to IfNotPresent, except for OCI images with tag latest, for which the default will be Always.
+
+|IfNotPresent
+|If an existing version of the image has been pulled before, that will be used. If no version of the image is present locally, we will pull the latest version.
+
+|Always
+|We will always pull the latest version of an image when applying this plugin.
+|===
+
+`Struct` represents a structured data value, consisting of fields which map to dynamically typed values. In some languages, Struct might be supported by a native representation. For example, in scripting languages like JS a struct is represented as an object. The details of that representation are described together with the proto support for the language.
```
Comment (jwendell) [inline modules/ossm-extensions-ref-wasmplugin.adoc:146]: Replace `JS` with `JavaScript`

Comment (jwendell) [reply modules/ossm-extensions-ref-wasmplugin.adoc:146]: Remove the final sentence `The details of that representation are described together with the proto support for the language.`

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+.Struct
```
Comment (jwendell) [inline modules/ossm-extensions-ref-wasmplugin.adoc:147]: This is missing a blank line I guess. The table got no title.

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+|===
+
+`PluginPhase` specifies the phase in the filter chain where the plugin will be injected.
+
+.PluginPhase
+[options="header"]
+[cols="a, a"]
+|===
+| Field | Description
+|UNSPECIFIED_PHASE
```
Comment (jwendell) [inline modules/ossm-extensions-ref-wasmplugin.adoc:164]: Replace this with `<empty>`

File: modules/ossm-extensions-smextension-deploy.adoc [hunk 12]
```diff
@@ -0,0 +1,42 @@
+////
```
Comment (jwendell) [inline modules/ossm-extensions-smextension-deploy.adoc:1]: Are you going to mention this is deprecated? Or should we just remove this section?

File: modules/ossm-extensions-wasmplugin-deploy.adoc [hunk 13]
```diff
@@ -0,0 +1,41 @@
+////
+This module included in the following assemblies:
+*service_mesh_/v2x/ossm-extensions.adoc
+////
+:_content-type: PROCEDURE
+[id="ossm-wasmplugin-deploy_{context}"]
+= Deploying `WasmPlugin` resources
+
+You can enable {SMProductName} extensions using the `ServiceMeshExtension` resource. In this example, `istio-system` is the name of the control plane project.  The following example creates an `openid-connect filter` that performs an OpenID Connect flow to authenticate the user.
```
Comment (jwendell) [inline modules/ossm-extensions-wasmplugin-deploy.adoc:9]: Replace `ServiceMeshExtension` with `WasmPlugin`

File: modules/ossm-extensions-wasmplugin-deploy.adoc [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+.Procedure
+
+. Create the following example resource:
++
+.Example ServiceMeshExtension resource extension.yaml
```
Comment (jwendell) [inline modules/ossm-extensions-wasmplugin-deploy.adoc:15]: You can use the filename `plugin.yaml` here and below if you wish. Not a strict requirement though.

File: modules/ossm-extensions-smextension-deploy.adoc (hunk 12 above)
Comment (JStickler) [reply modules/ossm-extensions-smextension-deploy.adoc:1]: There's a note up in the overview about the deprecation.  We're going to leave it for now.   We'll remove it once the API is officially no longer supported.

File: modules/ossm-extensions-wasmplugin-deploy.adoc (hunk 13 above)
Comment (JStickler) [reply modules/ossm-extensions-wasmplugin-deploy.adoc:9]: D'oh, copy/paste error!  Thanks for catching that.

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+      istio: ingressgateway
+  url: file:///opt/filters/openid.wasm
+  sha256: 1ef0c9a92b0420cf25f7fe5d481b231464bc88f486ca3b9c83ed5cc21d2f6210
+  phase: AUTHN
+  pluginConfig:
+    openid_server: authn
+    openid_realm: ingress
+----
+
+In the following example, an authentication filter is deployed to the ingress-gateway that implements an OpenID flow and populates the `Authorization` header with an Oracle Cloud Infrastructure (OCI) image.
```
Comment (jwendell) [inline modules/ossm-extensions-ref-wasmplugin.adoc:34]: Oracle is wrong here. But I'd suggest to rewrite this with something like:

Below is the same example, but this time an OCI image is used instead of a file in the filesystem. Note the fields `url`, `imagePullPolicy` and `imagePullSecret`.

File: modules/ossm-extensions-wasmplugin-deploy.adoc [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+. Create the following example resource:
++
+.Example plugin.yaml
+[source,yaml]
+----
+apiVersion: extensions.istio.io/v1alpha1
+kind: WasmPlugin
+metadata:
+  name: openid-connect
+  namespace: istio-ingress
```
Comment (jwendell) [inline modules/ossm-extensions-wasmplugin-deploy.adoc:22]: Replace with `istio-system`.

File: modules/ossm-extensions-ref-smextension.adoc [hunk 17]
```diff
@@ -0,0 +1,45 @@
+////
+This module included in the following assemblies:
+*service_mesh_/v2x/ossm-extensions.adoc
+////
+:_content-type: REFERENCE
+[id="ossm-wasm-ref-smextension_{context}"]
+= ServiceMeshExtention API reference
```
Comment (jwendell) [inline modules/ossm-extensions-ref-smextension.adoc:7]: ServiceMeshExtension

File: modules/ossm-extensions-ref-smextension.adoc [hunk 18]
```diff
@@ -0,0 +1,45 @@
+////
+This module included in the following assemblies:
+*service_mesh_/v2x/ossm-extensions.adoc
```
Comment (jwendell) [inline modules/ossm-extensions-ref-smextension.adoc:3]: I still think this section should start with a [big] note/warning that everything below this point is deprecated. When reading the docs, this looks like just another section that looks a bit duplicate from what we read above. It's a bit confusing.

If this is not possible, what about at least mention `[deprecated]` every time `ServiceMeshExtension` is mentioned in a section title from now on?

File: modules/ossm-extensions-ref-wasmplugin.adoc (hunk 15 above)
Comment (JStickler) [reply modules/ossm-extensions-ref-wasmplugin.adoc:34]: Bah, I looked up OCI and then wrote down the wrong definition for the acronym.   =(  It's Open Container Initiative.

File: modules/ossm-extensions-ref-smextension.adoc (hunk 18 above)
Comment (JStickler) [reply modules/ossm-extensions-ref-smextension.adoc:3]: Our doc tooling and standards have some restrictions around how I can handle this without breaking things.  I've added a note about deprecation to each of the ServiceMeshExtension topics.

**Review by jwendell — APPROVED** [review]

**Comment by skondkar** [comment]
Verified with the steps in preview doc. Created WasmPlugin for OSSM 2.2

When creating ServiceMeshExtenstion on 2.2, got the error for unrecognized type ServiceMeshExtension
Successfully created ServiceMeshExtenstion on 2.1 as per the example in the preview doc

Screenshot from 2022-06-01 16-19-57
Screenshot from 2022-06-01 16-28-25

**Comment by jwendell** [comment]
@skondkar Can you provide more details on the error on 2.2? Feel free to create a ticket in Jira with the details/step to reproduce and tag me there.

**Comment by skondkar** [comment]
@jwendell thanks,
Created the issue to track: https://issues.redhat.com/browse/OSSM-1597

**Comment by skondkar** [comment]
LGTM.
Tested the PR again after verifying issue [OSSM-1597] on OSSM 2.2 latest build.
Created WasmPlugin resources and ServiceMeshExtension resource successfully on SMCP v2.2 as per the preview doc.

Verified that while creating ServiceMeshExtension, it shows below warning:
Warning: ServiceMeshExtension is deprecated. It works now but will be removed in a future version. Please use WasmPlugin instead.

**Comment by JStickler** [comment]
@jwendell @skondkar if I'm reading the comments on this correctly, now that we have a user friendly message about the ServiceMeshExtension, am I good to move this PR to peer review?  Or is there something else I need to update?

**Comment by JStickler** [comment]
Received ack via Slack that this PR is good to go and can move to peer review.

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+.PluginPhase
+[options="header"]
+[cols="a, a"]
+|===
+| Field | Description
+|<empty>
+|Control plane decides where to insert the plugin. This will generally be at the end of the filter chain, right before the Router. Do not specify PluginPhase if the plugin is independent of others.
+
+|AUTHN
+|Insert plugin before Istio authentication filters.
```
Comment (jldohmann) [inline modules/ossm-extensions-ref-wasmplugin.adoc:169]:
```suggestion
|Insert the plug-in before Istio authentication filters.
```

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+
+|AUTHZ
+|Insert plugin before Istio authorization filters and after Istio authentication filters.
```
Comment (jldohmann) [inline modules/ossm-extensions-ref-wasmplugin.adoc:172]:
```suggestion
|Insert the plug-in before Istio authorization filters and after Istio authentication filters.
```

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+
+|STATS
+|Insert plugin before Istio stats filters and after Istio authorization filters.
```
Comment (jldohmann) [inline modules/ossm-extensions-ref-wasmplugin.adoc:175]:
```suggestion
|Insert the plug-in before Istio stats filters and after Istio authorization filters.
```

File: modules/ossm-extensions-smextension-deploy.adoc [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+= Deploying `ServiceMeshExtension` resources
+
+You can enable {SMProductName} extensions using the `ServiceMeshExtension` resource. In this example, `istio-system` is the name of the control plane project.
+
+[NOTE]
+====
+When creating new WebAssembly extensions, use WasmPlugin. ServiceMeshExtension is deprecated as of {SMProductName} version 2.2 and will be removed in a future release.
+====
+
+For a complete example that was built using the Rust SDK, take a look at the link:https://github.com/maistra/header-append-filter[header-append-filter]. It is a simple filter that appends one or more headers to the HTTP responses, with their names and values taken out from the `config` field of the extension. See a sample configuration in the snippet below.
```
Comment (jldohmann) [inline modules/ossm-extensions-smextension-deploy.adoc:16]: ISG nits:
```suggestion
For a complete example that is built using the Rust SDK, take a look at the link:https://github.com/maistra/header-append-filter[header-append-filter]. It is a filter that appends one or more headers to the HTTP responses, with their names and values taken out from the `config` field of the extension. See a sample configuration in the snippet below.
```

File: modules/ossm-extensions-smextension-format.adoc [hunk 23]
```diff
@@ -0,0 +1,62 @@
+////
+This module included in the following assemblies:
+*service_mesh_/v2x/ossm-extensions.adoc
+////
+:_content-type: REFERENCE
+[id="ossm-extensions-smextension-format_{context}"]
+= `ServiceMeshExtension` container format
+
+You must have a `.wasm` file containing the bytecode of your WebAssembly module, and a `manifest.yaml` file in the root of the container filesystem to make your container image a valid extension image.
```
Comment (jldohmann) [inline modules/ossm-extensions-smextension-format.adoc:9]: Though I didn't find this in ISG, I think there is more usage of "file system" in the docs than "filesystem", so this is just a suggestion for consistency. I leave the decision up to you 🙂 
```suggestion
You must have a `.wasm` file containing the bytecode of your WebAssembly module, and a `manifest.yaml` file in the root of the container file system to make your container image a valid extension image.
```

File: modules/ossm-extensions-smextension-format.adoc [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+phase: PreAuthZ
+priority: 100
+module: extension.wasm
+----
+
+.Field Reference for manifest.yml
+[options="header"]
+[cols="a, a, a"]
+|===
+| Field | Description |Required
```
Comment (jldohmann) [inline modules/ossm-extensions-smextension-format.adoc:33]: For consistency (and simplicity), a suggestion that the "Required Fields" column just be "Yes" or "No"

File: modules/ossm-extensions-wasmplugin-deploy.adoc [hunk 25]
```diff
@@ -0,0 +1,41 @@
+////
+This module included in the following assemblies:
+*service_mesh_/v2x/ossm-extensions.adoc
+////
+:_content-type: PROCEDURE
+[id="ossm-wasmplugin-deploy_{context}"]
+= Deploying `WasmPlugin` resources
+
+You can enable {SMProductName} extensions using the `WasmPlugin` resource. In this example, `istio-system` is the name of the control plane project.  The following example creates an `openid-connect filter` that performs an OpenID Connect flow to authenticate the user.
```
Comment (jldohmann) [inline modules/ossm-extensions-wasmplugin-deploy.adoc:9]: I'm unsure if `filter` is supposed to be monospaced here, because it doesn't match the what is in the YAML code block. Just disregard if it's correct 🙂 
```suggestion
You can enable {SMProductName} extensions using the `WasmPlugin` resource. In this example, `istio-system` is the name of the control plane project.  The following example creates an `openid-connect` filter that performs an OpenID Connect flow to authenticate the user.
```

File: modules/ossm-extensions-wasmplugin-format.adoc [hunk 26]
```diff
@@ -0,0 +1,15 @@
+////
+This module included in the following assemblies:
+*service_mesh_/v2x/ossm-extensions.adoc
+////
+:_content-type: REFERENCE
+[id="ossm-extensions-wasmplugin-format_{context}"]
+= `WasmPlugin` container format
+
+Istio supports Open Container Initiative (OCI) images in its Wasm Plugin mechanism. You can distribute your Wasm Plugins in the form of a container image, and you can use `spec.url` to refer to a container registry location.  For example, `quay.io/my-username/my-plugin:latest`.
```
Comment (jldohmann) [inline modules/ossm-extensions-wasmplugin-format.adoc:9]: API object reference and minor suggestion:
```suggestion
Istio supports Open Container Initiative (OCI) images in its Wasm Plugin mechanism. You can distribute your Wasm Plugins as a container image, and you can use the `spec.url` field to refer to a container registry location.  For example, `quay.io/my-username/my-plugin:latest`.
```

File: modules/ossm-extensions-wasmplugin-format.adoc [hunk 27]
```diff
@@ -0,0 +1,15 @@
...
+
+Because each execution environment (runtime) for a WASM module can have runtime-specific configuration parameters, a WASM image can be composed of two layers:
+
+* *plugin.wasm* (Required) - Content layer. This layer consists of a `.wasm` binary containing the bytecode of your WebAssembly module, to be loaded by the runtime.  This file must be named `plugin.wasm`.
```
Comment (jldohmann) [inline modules/ossm-extensions-wasmplugin-format.adoc:13]: Minor nit for active voice:
```suggestion
* *plugin.wasm* (Required) - Content layer. This layer consists of a `.wasm` binary containing the bytecode of your WebAssembly module, to be loaded by the runtime.  You must name this file `plugin.wasm`.
```

File: modules/ossm-extensions-wasmplugin-format.adoc [hunk 28]
```diff
@@ -0,0 +1,15 @@
...
+
+* *runtime-config.json* (Optional) - Configuration layer. This layer consists of a JSON-formatted string that describes metadata describing the module for the target runtime. The config layer may also contain additional data, depending on the target runtime. For example, the config for a WASM Envoy Filter contains root_ids available on the filter.
```
Comment (jldohmann) [inline modules/ossm-extensions-wasmplugin-format.adoc:15]: ISG nit:
```suggestion
* *runtime-config.json* (Optional) - Configuration layer. This layer consists of a JSON-formatted string that describes metadata about the module for the target runtime. The config layer might also contain additional data, depending on the target runtime. For example, the config for a WASM Envoy Filter contains root_ids available on the filter.
```

**Review by jldohmann — COMMENTED** [review]
Looking good! Mostly some minor ISG nits and API object references. I realize that a lot of what GH perceives as file changes in content are actually from modularization, so if any of this is out of scope with your changes, I apologize

File: modules/ossm-extensions-ref-wasmplugin.adoc [hunk 29]
```diff
@@ -0,0 +1,176 @@
...
+`WorkloadSelector` specifies the criteria used to determine if a filter can be applied to a proxy. The matching criteria includes the metadata associated with a proxy, workload instance information such as labels attached to the pod/VM, or any other information that the proxy provides to Istio during the initial handshake. If multiple conditions are specified, all conditions need to match in order for the workload instance to be selected. Currently, only label based selection mechanism is supported.
+
+.WorkloadSelector
+[options="header"]
+[cols="a, a, a, a"]
+|===
+| Field | Type | Description | Required
+|matchLabels
+|map<string, string>
+|One or more labels that indicate a specific set of pods/VMs on which a policy should be applied. The scope of label search is restricted to the configuration namespace in which the resource is present.
```
[the remark this answers is not in the document]
Comment (JStickler) [reply modules/ossm-extensions-ref-wasmplugin.adoc:122]: Um, now two sentences end with prepositions?  Julie is confused.

[the remark this answers is not in the document]
Comment (jldohmann) [reply modules/ossm-extensions-ref-wasmplugin.adoc:122]: @JStickler my bad, I didn't explain that well. ISG says to _not_ avoid ending sentences in prepositions...so ending them in prepositions is preferred. I wish the ISG would've just said that 😅

**Comment by JStickler** [comment]
Added note that WebAssembly extensions not supported on IBM Power and Z systems at request of IBM technical writers.

**Comment by JStickler** [comment]
Merging now to get this into main so that Eng and 3scale team can work on documenting the next steps.
